src/field/vecops.cxx

```diff
@@ -37,76 +37,97 @@
 const Vector2D Grad(const Field2D &f, CELL_LOC outloc) {
   TRACE("Grad( Field2D )");
 
+  CELL_LOC outloc_x, outloc_y, outloc_z;
+  if (outloc == CELL_VSHIFT) {
+    outloc_x = CELL_XLOW;
+    outloc_y = CELL_YLOW;
+    outloc_z = CELL_ZLOW;
+  } else {
+    outloc_x = outloc_y = outloc_z = outloc;
+  }
+
   Vector2D result(f.getMesh());
 
-  result.x = DDX(f, outloc);
-  result.y = DDY(f, outloc);
-  result.z = DDZ(f, outloc);
+  result.x = DDX(f, outloc_x);
+  result.y = DDY(f, outloc_y);
+  result.z = DDZ(f, outloc_z);
 
-  result.setLocation(outloc);
+  if (outloc == CELL_DEFAULT) {
+    result.setLocation(result.x.getLocation());
+  } else {
+    result.setLocation(outloc);
+  }
 
   result.covariant = true;
 
   return result;
 }
 
-const Vector3D Grad(const Field3D &f, CELL_LOC outloc_x, CELL_LOC outloc_y,
-                    CELL_LOC outloc_z) {
+const Vector3D DEPRECATED(Grad(const Field3D &f, CELL_LOC outloc_x, CELL_LOC outloc_y,
```

Comment:
This should go on the declaration in the header rather than the implementation.

Also, the single-arg version needs to go in the header. Note that I think you'll need to remove the default values for the three-arg version and add one for the single-arg.

Reply:
Ah, forgot the header.

Should I remove deprecated from the source file and just put it on header?

Reply:
Also doesn't removing the default values break backwards compatibility?